Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs

```diff
@@ -0,0 +1,210 @@
+﻿using System;
+using System.Composition.Hosting.Core;
+using System.Linq;
+using System.Net.Security;
+using Content.Server.GameObjects.Components.Atmos;
+using Content.Server.GameObjects.Components.Body.Circulatory;
+using Content.Server.GameObjects.Components.Body.Respiratory;
+using Content.Server.Utility;
+using Content.Shared.Chemistry;
+using Content.Shared.Interfaces.GameObjects.Components;
+using Robust.Server.GameObjects;
+using Robust.Shared.GameObjects;
+using Robust.Shared.GameObjects.Components.Timers;
+using Robust.Shared.GameObjects.Components.Transform;
+using Robust.Shared.Interfaces.GameObjects;
+using Robust.Shared.Interfaces.Map;
+using Robust.Shared.IoC;
+using Robust.Shared.Maths;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization;
+
+namespace Content.Server.GameObjects.Components.Chemistry
+{
+    [RegisterComponent]
+    public class SmokeComponent : Component
+    {
+        private const float ExposeDelay = 0.5f;
+
+        public override string Name => "Smoke";
+
+        [Dependency] private readonly IMapManager _mapManager = default!;
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+        private IEntity _inception;
+
+        private bool _running;
+        private float _timer;
+        private float _exposeTimer;
+
+        private int _amount;
+        private float _lifetime;
+        private float _spreadDelay;
+        private float _averageLifetime;
+
+        public bool IsInception => _inception == Owner;
+
+        public void Activate(Solution solution, int amount, float duration, float spreadDelay)
+        {
+            if (_running)
+                return;
+            TryAddSolution(solution);
+            Start(Owner, amount, duration+2*amount*spreadDelay, spreadDelay, duration+amount*spreadDelay);
+        }
+
+        public void Update(float frameTime)
+        {
+            if (!_running)
+                return;
+
+            if (IsInception)
+            {
+                if (_exposeTimer > ExposeDelay)
+                {
+                    _exposeTimer = 0;
+                    ReactWithTileAndEntities();
+                }
+                _exposeTimer += frameTime;
+            }
+            else
+            {
+                if (_inception.TryGetComponent(out SmokeComponent smokeComp) &&
+                    smokeComp._exposeTimer > ExposeDelay)
+                {
+                    ReactWithTileAndEntities();
+                }
+            }
```

Comment:
Rather than chucking all of this into the component IMO you're better off having the system keep track of which components are connected which should also have better perf.